Remove shipping if there is no address, always allow hide shipping costs option

samueljseay opened the pull request:
Closes #52335 

With some adjustments:

1. We used to disable "hide shipping costs until an address is entered" when local pickup was enabled. We now allow it to be enabled but we explain the behaviour (`Local pickup rates will display in the Checkout block, even without an address`). @opr and I agreed this experience felt clearer and covered all cases well, but it would be good to get some feedback about this cc @pmcpinto 

2. If an address has not been entered and the user has hide shipping costs enabled we do not display the *Shipping* options

Added by @opr 

- Changes the shipping calculator logic to display "Enter an address to check delivery options" if only a partial address has been entered (e.g. when the shopper has a default location)
- We previously hid the "Ship" selector when:
  - Shipping is not required
  - Shipping calculations have not been completed
  - No shipping rates are available
  - The item is not collectible
  - Local pickup is not enabled
  - No shipping methods are currently available
- however this would hide it even if rates _might_ become available later, after a full address had been entered. Instead I changed the logic to force the control to show if the following is true:
  - Shipping is enabled, and
  - The cart needs shipping, and
  - All packages in the cart support collectible, and
  - Local pickup is enabled, and
  - Shipping methods exist on the site
- Updated shipping calculator so it doesn't show a "No delivery options available for XXX, XX" message unless there's a full address. I figured the shopper could be negatively affected by such a message if they haven't even filled their address in and they're already seeing a message saying there are no shipping methods. Because of this, I updated the tests in `plugins/woocommerce-blocks/tests/e2e/tests/cart/cart-checkout-block-shipping.block_theme.spec.ts`.
- Added a `has_full_shipping_address` method to `ShippingController` - this uses `WC()->countries->get_country_locale` to determine whether the state, city, country, postcode (based on locale requirement) is filled before stripping out "shipping" methods (and leaving pickup methods intact).
  - Needed because we have to determine whether to remove rates from the response if shipping requires a full address.
  

### How to test the changes in this Pull Request:

Smoke test a few scenarios to ensure no regressions. For the shopper part it is best to open an incognito window to clear the address state that would be remembered between scenarios.

#### Scenario 1
1. Have local pickup disabled
2. Have hide shipping costs until address entered **enabled**
3. As shopper go through cart and checkout and ensure that shipping costs are not shown on cart page / checkout page until address entered

#### Scenario 2

1.  Have local pickup disabled
2. Have hide shipping costs until address entered **disabled**
3. As a shopper go through cart and checkout and ensure that shipping and pick up options are visible before address is entered

#### Scenario 3

1. Have local pickup enabled
2. Have hide shipping costs until address entered **enabled**
3.  As a shopper go through cart and checkout and ensure that shipping options are not shown but pickup options are still shown.

#### Scenario 4

1. Go to WC -> Settings -> General. Set Shipping location(s) to `Disable shipping & shipping calculations`
2. Try to check out with a physical product. You should not be able to enter a shipping address or select local pickup. Checkout should succeed.
3. Try to check out with a virtual product. You should not be able to enter a shipping address or select local pickup. Checkout should succeed.

#### General testing

Test with various setups of:
- Local pickup enabled/disabled
- Hide shipping costs until an address is entered enabled/disabled
- Shipping calculator enabled/disabled
- No shipping zones enabled on the whole site
- Only fallback shipping zones enabled
- No fallback zones enabled, but methods exist for specific countries. Test with addresses in and out of these countries.
- Mixtures of virtual, downloadable, and physical products in the cart. Try with each type individually too.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Improve how we show/hide shipping address at cart and checkout based on options.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

samueljseay committed:
Remove shipping if there is no address, always allow hide shipping costs option.

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

samueljseay commented:
Hey @pmcpinto,

@opr and I did some adjustments of how we handle hiding/showing the addresses for shipping and pick up in this PR with a slight modification to the behaviour and wording of the `Hide shipping costs until an address is entered`. I was wondering if you could have a look at the new behaviour and see if it's ok.

pmcpinto commented:
> @opr and I did some adjustments of how we handle hiding/showing the addresses for shipping and pick up in this PR with a slight modification to the behaviour and wording of the Hide shipping costs until an address is entered. I was wondering if you could have a look at the new behaviour and see if it's ok.

Thanks for the ping. The changes look good to me.

opr commented on plugins/woocommerce/src/Blocks/Shipping/ShippingController.php:
```diff
@@ -433,6 +430,42 @@ function ( $rate ) {
 	}
 
+	/**
+	 * Remove shipping if no address is entered.
```
Even though we just paired on this, it took me a minute to remember what this was for and figure out why. I think the comment could be updated, I made a suggestion:

```suggestion
	 * Remove shipping (i.e. delivery, not local pickup) if "Hide shipping costs until an address is entered" is enabled, and no address has been entered yet.
```

opr commented on plugins/woocommerce/src/Blocks/Shipping/ShippingController.php:
```diff
@@ -433,6 +430,42 @@ function ( $rate ) {
+	 * @return array
+	 */
+	public function remove_shipping_if_no_address( $packages ) {
+		$customer = wc()->customer;
```
Lint error here, equals not lined up

opr commented on plugins/woocommerce/src/Blocks/Shipping/ShippingController.php:
```diff
@@ -433,6 +430,42 @@ function ( $rate ) {
+		if ( ! $has_address && $option_checked ) {
+			$packages = array_map(
+				function ( $package ) {
+					if ( ! is_array( $package['rates'] ) ) {
```
Do we want to check if `$package['rates']` is set first before we try is_array on it, just thinking if $package doesn't have a member called rates this will warn, won't it?

opr commented on plugins/woocommerce/src/Blocks/Shipping/ShippingController.php:
```diff
@@ -433,6 +430,42 @@ function ( $rate ) {
+					$package['rates'] = array_filter(
+						$package['rates'],
+						function ( $rate ) {
+							return in_array( $rate->get_method_id(), LocalPickupUtils::get_local_pickup_method_ids(), true );
```
Do we want to check if `$rate instanceof WC_Shipping_Rate` just in case?

samueljseay replied:
Can't hurt, but I'm curious what would make us be so defensive of this? Shouldn't we rather see fatal error so we can fix if we ever did not have a WC_Shipping_Rate in this array?

samueljseay replied:
Oh right because this comes from a filter ofc. Yeah I'll add the defensive check.

samueljseay replied:
Resolved in https://github.com/woocommerce/woocommerce/pull/53466/commits/7c5344bbdec33a7a4523cd4cdf6b13527cf0ecd9

opr reviewed (commented):
Change makes sense to me, doesn't _solve_ the original issue, but the behaviour is improved and clearer isn't it?

I think you should ask for a second review since I helped you write this, however I made a couple of comments anyway that you may want to look at.

mikejolley commented on plugins/woocommerce/src/Blocks/Shipping/ShippingController.php:
```diff
@@ -142,13 +144,8 @@ public function remove_shipping_settings( $settings ) {
-					);
-					$settings[ $index ]['disabled'] = true;
-					$settings[ $index ]['value']    = 'no';
+					$settings[ $index ]['desc'] =
```
Look at how "Debug mode" is presented on the same screen with the extra description below. I would add the second part like that to keep the label short.

`Local pickup rates will display in the Checkout block, even without an address`

mikejolley replied:
This part is missing in test instructions FYI.

samueljseay replied:
Great suggestion thanks. Fixed in https://github.com/woocommerce/woocommerce/pull/53466/commits/3d9aae0f98dba5cda4ba9c29c05dcc7d53ffcceb

mikejolley commented on plugins/woocommerce/src/Blocks/Shipping/ShippingController.php:
```diff
@@ -433,6 +430,42 @@ function ( $rate ) {
+		$has_address = $customer->has_shipping_address();
+
+		remove_filter( 'option_woocommerce_shipping_cost_requires_address', array( $this, 'override_cost_requires_address_option' ) );		
+		$option_checked = filter_var( get_option( 'woocommerce_shipping_cost_requires_address', 'no' ), FILTER_VALIDATE_BOOLEAN );
```
I don't know which is "better" but there is a `wc_string_to_bool` function which does this.

samueljseay replied:
I find filter var and it's wording about "validating a boolean" doesnt really match the idea of a straight conversion, so yeah I'm happy to use the woo one here.

samueljseay replied:
Resolved in https://github.com/woocommerce/woocommerce/pull/53466/commits/ba1ce0874440608568fc0b2107bb0bf7f36cfeaa

mikejolley commented on plugins/woocommerce/src/Blocks/Shipping/ShippingController.php:
```diff
@@ -89,9 +90,10 @@ function () {
 	 */
 	public function override_cost_requires_address_option( $value ) {
-		if ( CartCheckoutUtils::is_checkout_block_default() && $this->local_pickup_enabled ) {
+		if ( is_checkout() && ! is_admin() && CartCheckoutUtils::is_checkout_block_default() && $this->local_pickup_enabled ) {
```
Why is the admin check needed when there is a check for `is_checkout()`?

samueljseay replied:
Thats a good point. It will never be admin if is_checkout is true. I'll change this and retest. Thanks for spotting this.

samueljseay replied:
Resolved in https://github.com/woocommerce/woocommerce/pull/53466/commits/ba1ce0874440608568fc0b2107bb0bf7f36cfeaa

mikejolley reviewed (commented):
Seems to work fine. I added some comments in addition to Thomas' feedback.

samueljseay committed:
Fix alignment issues via autoformatting, update comment to better describe intent

samueljseay committed:
Use desc_tip to display additional details below the setting description.

samueljseay committed:
Use wc_string_to_bool instead of filter_var, remove redundant is_admin from logic.

samueljseay committed:
Add extra defensive check for instance of WC_Shipping_Rate

samueljseay commented:
@opr @mikejolley Thanks so much for all the great suggestions. I've implemented and re-tested. Would you mind taking another look? 🙇

mikejolley commented:
I'm seeing some very strange behaviour on this branch; maybe @opr can confirm.

1. Hide shipping options **enabled**
2. Local pickup options **enabled**
3. Incognito window. Add to cart. 
4. Cart page shows no shipping or pickup options
5. Checkout page shows pickup options. Change the pickup option. _Shipping address appears and pickup is hidden. Shipping/collection toggle block is also hidden_

This seems to be a bug.

opr reviewed (commented):
Can also reproduce what Mike is seeing - Scenarios 1 and 2 are working OK but 3 is buggy.

Looks like [this line](https://github.com/woocommerce/woocommerce/blob/d1498d45f2804eede99538d13717c4e2cf602308/plugins/woocommerce/src/StoreApi/Schemas/V1/CartSchema.php#L346) is the cause, when selecting a local pickup method, the cart response is re-fetched, but because no address data is entered the packages are not returned in the response. (Didn't dig into why they are returned at first, though).

If you update that line to:

```php
		$shipping_packages = $has_calculated_shipping || LocalPickupUtils::is_local_pickup_enabled() ? $controller->get_shipping_packages() : [];
```

it seems to work, not sure if that's the best fix though.

samueljseay commented:
Just noting that I've talked to @opr and will hand this PR over to him to take over the line while I focus on other priorities. I'm moving this PR into draft and then he can decide what to do next with it.

opr committed:
Prevent hiding the shipping/pickup selector if an address is required

There might be rates to discover yet

opr committed:
Add has_full_shipping_address to ShippingController

opr committed:
Always calculate shipping packages in Cart response

In case there is local pickup

opr committed:
Check customer has full address before filtering out regular methods

opr committed:
Remove address_1 from full address requirement

opr committed:
Hide delivery rates until address entered if setting is on

opr committed:
Fix lint issues

opr committed:
Update comment for has_full_shipping_address

opr committed:
Bust locale cache so the filter can be applied in checkout test

opr committed:
Rename var and optimize so fewer functions called

opr committed:
Rework shipping method selector logic

opr committed:
Update test for shipping calculator text

opr committed:
Update test to verify shipping selector doesn't show if no methods

opr committed:
Update local pickup settings to use noticesStore

opr committed:
Update local pickup test to use noticesStore

opr committed:
Add test to ensure ship/pickup toggle doesn't show if no methods on site

opr committed:
Update comments

opr commented on plugins/woocommerce/src/StoreApi/Schemas/V1/CartSchema.php:
```diff
@@ -342,8 +343,12 @@ public function get_item_response( $cart ) {
+		// Get shipping packages to return in the response from the cart. Always get the shipping packages if local
+		// pickup is enabled and has methods. If the address is required then regular shipping rates will be filtered
+		// out later.
+		$shipping_packages = $has_calculated_shipping || $has_local_pickup_methods ? $controller->get_shipping_packages() : [];
```
If there are local pickup methods, we should return these regardless of whether an address has been entered. `$has_calculated_shipping` is a _little_ misleading, as `$cart->show_shipping()` checks whether shipping is enabled and cart has items, and if an address is required to show shipping, returns whether the address fields are complete, otherwise returns true, so it's not really to do with calcuation completeness, but address completeness.

wavvves replied:
Why not rename it and update comments? maybe `$can_calculate_shipping`.
We have other bits that can get confusing when looking at the var name, like `WC()->customer->has_calculated_shipping()`

wavvves replied:
~Also, by definition this sounds like  `ShippingController::has_full_shipping_address()` introduced in this PR~
Nevermind, the other goes beyond that to check if shipping can appear in Totals.
I see there's an issue to clear this https://github.com/woocommerce/woocommerce/issues/54894 :)

opr committed:
Add SHIPPING_ENABLED const to read enabled/disabled option

opr committed:
Don't show Ship/Pickup selector if shipping is disabled

opr commented on plugins/woocommerce/tests/php/src/Blocks/StoreApi/Routes/Checkout.php:
```diff
@@ -599,6 +599,7 @@ function ( $locale ) {
 			}
 		);
 
+		unset( WC()->countries->locale );
```
This is needed so the filter applied above actually works. This filter is applied after the locale is cached.

opr committed:
Clean up and update logic in has_full_address

opr committed:
Add unit test for has_full_address

opr committed:
Update false negative in E2E tests

opr committed:
Add new E2E tests to cover a couple of new scenarios

opr committed:
Fix lint issues

opr committed:
Move settings to core

Not needed because nothing changes besides the tooltip which is true in both shortcode and blocks

opr committed:
Update logic to show shipping method selector

Now it will only show if shipping is enabled, cart needs shipping, local pickup is enabled, and some methods are enabled, and the cart is collectible

opr committed:
Remove override on woocommerce_shipping_cost_requires_address

We can just use the setting in the database now because respect core settings, previously it was disabled in some cases to make sure we could calculate shipping costs

opr committed:
Force rates back to an array if it was filtered to something else

opr committed:
Revert "Move settings to core"

This reverts commit bf282f8b99175b04a4b3767b456aeb29d244788d.

opr committed:
Remove setting that is the same as core

opr committed:
Update comment

opr committed:
Remove redundant type check

opr committed:
Test with default false, country-specific true

opr committed:
Prevent bug if default is false and country-specific is true

opr committed:
Remove unused var

opr committed:
Remove unused import

opr committed:
Add test to ensure shipping doesn't show for virtual products

opr committed:
Update tests with false negatives

opr committed:
Don't show shipping calculator if no methods exist

opr committed:
Add util to make outputting shippingMethodsExist easier

opr committed:
Updare Cart and Checkout blocks to add shippingMethodsExist to registry

opr committed:
Add test for when shipping calculator is disabled

mikejolley commented on plugins/woocommerce/src/Blocks/Shipping/ShippingController.php:
```diff
@@ -433,6 +412,82 @@ function ( $rate ) {
+	 * Only required fields are checked.
+	 */
+	public function has_full_shipping_address() {
+		$customer = WC()->customer;
```
Unlikely but to be defensive I would include a check to ensure this is a `WC_Customer` object. Before cart init this can be null.

opr replied:
Thanks, I think it's fine for this method to return false if that's the case, since the rates are not going to show anywhere anyway. Done in 1472929c07b9bdfc1e7206ee3dce63a3518534ea

mikejolley commented on plugins/woocommerce/src/Blocks/Shipping/ShippingController.php:
```diff
@@ -433,6 +412,82 @@ function ( $rate ) {
+		$country_locale   = $address_fields[ $locale_key ] ?? array();
+
+		// For all fields in $shipping_address, check if they are required in the country-specific locale first, if
+		// not set, orn$address_fields and if so, check if they are not empty.
```
`orn$address_fields`?

opr replied:
Not sure where that came from, but re-wrote this comment in 893a7b9c28798aa5f01756256336fee58120503a

mikejolley commented on plugins/woocommerce/src/Blocks/Shipping/ShippingController.php:
```diff
@@ -433,6 +412,82 @@ function ( $rate ) {
+
+		$has_full_address = $this->has_full_shipping_address();
+
+		if ( ! $has_full_address ) {
```
We should tweak this to return early. So have:

```php
if ( $has_full_address ) {
	return $packages;
}
```

before the array map.

opr replied:
Thanks, done in 65f8bc8601610b8631312899fcd0e29f904d1fca

mikejolley reviewed (commented):
Minor feedback attached. I tested a variety of flows. One thing stuck out as odd and it was this case.

1. Hide rates until address entered
2. Shop base UK. Local Pickup enabled. Local pickup location something odd like US address.
3. Go to cart. It says enter an address to see rates.
4. Enter UK address in shipping calculator. Submit.
5. Rates appear but the address shows a US address (the pickup location)

Its not a bug per-se but I had a moment where I was confused because I entered a UK address but it said I was collecting from a US address. Obviously I needed to pick a rate, but maybe it needs to reset selection after populating rates?

@opr

opr committed:
Return early if an address is entered

opr committed:
Clarify comment on getting required from locale loop

opr committed:
Check $customer is an instance of WC_Customer before operating on it

senadir commented on plugins/woocommerce/includes/class-wc-shipping.php:
```diff
@@ -367,6 +367,12 @@ public function calculate_shipping_for_package( $package = array(), $package_key
 			 */
 			$package['rates'] = apply_filters( 'woocommerce_package_rates', $package['rates'], $package );
 
+			// Package rates should be an array, if it was filtered into a non-array, reset it. Don't reset to the
+			// unfiltered value, as e.g. a 3pd could have set it to "false" to remove rates.
+			if ( ! is_array( $package['rates'] ) ) {
+				$package['rates'] = array();
+			}
+
```
Did you actually run into an issue here?

opr replied:
Well, no not any specific issue, but we need to make sure it's an array before continuing. I originally wanted to just reset to the original rates, but Mike mentioned 

> Don't reset to the unfiltered value, as e.g. a 3pd could have set it to "false" to remove rates.

It seems fine to just empty it, because if they're actually changing the rates correctly, an array would be returned, I can't think of any use case where rates would be modified into a non-array and expected to show anything.

senadir reviewed (approved):
Reviewed the code changes and they make sense to me, nothing that looks off.

alexflorisca reviewed (approved):
I haven't reviewed the code, but I did go through the testing instructions and it all looks pretty groovy. I tried pretty hard to break it, setting the shop to only sell and ship to specific countries, and it works well. Just one thing probably worth mentioning.

When the shop is restricted to selling to only select countries, local pickup locations that are **not** in those countries are shown in the list. E.g. If I only sell to UK and US, and I have a pickup location in France, this shows in the list. I don't think this is a huge problem, as it's more of a setup problem by the merchant, but wondering if we should hide local pickup locations that are outside of the countries the shop sells to?

opr commented:
> When the shop is restricted to selling to only select countries, local pickup locations that are **not** in those countries are shown in the list. E.g. If I only sell to UK and US, and I have a pickup location in France, this shows in the list. I don't think this is a huge problem, as it's more of a setup problem by the merchant, but wondering if we should hide local pickup locations that are outside of the countries the shop sells to?

Interesting use case, I think the distinction here is that local pickup is a _shipping_ address, so the customer may be billed in UK, but send their friend to collect in France in your above example. It seems weird and is probably not likely to ever be set up like that, so I'm happy to leave this as it is.

There is a bug (not caused by this PR) where if you have local pickup options in different countries, then set values in "Ship to specific countries only" then no local pickup options appear unless _all_ local pickup options are in a supported country.

Think that's for another day though the scope of this PR shouldn't grow any bigger!

wavvves reviewed (approved):
Test steps, I messed around with it but didn't spot anything.
Approving, and left a nitpick.
